feat: complete identify protocol and improved tests

acul71 opened the pull request:
**NOTE: this is a draft pull request**
The returned info agent version was a generic `py-libp2p/unknown`
I've added a function in `libp2p/utils.py` that return the agent version so that I can return it in the `agentVersion` field of `identify protocol` modifying `libp2p/identity/identify/protocol.py`

I've improved the test `tests/core/identity/identify/test_protocol.py` adding tests so that every field is verified.

I'm planning to complete https://github.com/libp2p/py-libp2p/issues/358 adding missing `observed address` from `stream` and then `identify/push` .
Specs are here https://github.com/libp2p/specs/blob/master/identify/README.md 

🕷️

acul71 committed:
feat: add agent version to identify protocol and improved tests

pacrob commented:
@acul71, I'm unclear if this is ready for review. It will need a rebase and a newsfragment, otherwise looking good so far.

acul71 commented:
> @acul71, I'm unclear if this is ready for review. It will need a rebase and a newsfragment, otherwise looking good so far.

This code is ready for review.
But does not include all features requested by identify protocol.
Anyway Manusheel suggested me to commit PR of my work on identify little by little, so that I can get feedback, instead of a "big" commit at the end of the work

seetadev commented:
@acul71 : Hi Luca. Kindly add a note that this a draft pull request. I will keep reviewing side by side.

For final review by Paul, I will leave a message tagged to him.

@pacrob : Wish to share that I recommended Luca and other PLDG cohort members to share the pull request with unit and integration tests. Also, step by step so that we could discuss their work in progress during the maintainer's call every Thursday.

> @pacrob
> 
> > @acul71, I'm unclear if this is ready for review. It will need a rebase and a newsfragment, otherwise looking good so far.
> 
> This code is ready for review. But does not include all features requested by identify protocol. Anyway Manusheel suggested me to commit PR of my work on identify little by little, so that I can get feedback, instead of a "big" commit at the end of the work PS: doing make lint, it did lint the README.md file (not sure why) so I commit also this, even if it's unrelated

acul71 committed:
feat: add observed_addr to identify protocol

acul71 commented:
@pacrob @seetadev 
With this commit https://github.com/libp2p/py-libp2p/pull/506/commits/f24c6c85622d1284e922776082989aa5ea75f5c4 the identify protocol is complete including all the requested fields.
In particular I've added the missing `observed_addr` to `identify` protocol
I saw that `get_peerstore` method was removed from `IHost` interface. Not sure why but I added it back (It's implemented in class `BasicHost(IHost)` )
I've added a test so now `it should be able to identify another peer`
Planning to add this other tests:
- `it('should register for identify protocol on start')`
- `it('should time out during identify')`

🕷️

acul71 committed:
feat: implement get_remote_address via delegation pattern

acul71 committed:
feat: Add observed addresses of remote peer using get_remote_address

acul71 commented:
@pacrob @seetadev 
# Identify Protocol Enhancement: Remote Peer Observed Address

## Overview  
This PR enhances the identify protocol to include the observed address of remote peers. When a node receives an identify request, it now determines how it sees the connecting peer (the observed address) and includes this information in the identify response.  

## Motivation  
In peer-to-peer networks, NAT traversal and correct address advertisement are critical challenges. Peers often need to know how they're seen by others on the network to:  

- Discover their public IP/port when behind NAT  
- Validate that advertised addresses are reachable  
- Improve connection reliability by advertising working addresses  

This implementation matches the behavior in other libp2p implementations, allowing Python-libp2p to fully participate in the network's address discovery process.  

## Technical Implementation  

### Changes Made  
- Enhanced the identify handler to extract the remote address from incoming connections  
- Added conversion from socket address tuples to libp2p multiaddresses  
- Included the observed address in the identify protocol response  
- Fixed various interface implementations in the connection chain to support remote address delegation  

### Code Walkthrough  
The identify protocol handler now includes logic to retrieve and report observed addresses:  

```python
def identify_handler_for(host: IHost) -> StreamHandlerFn:
    async def handle_identify(stream: INetStream) -> None:
        peer_id = stream.muxed_conn.peer_id

        # Get the remote address
        try:
            remote_address = stream.get_remote_address()
            # Convert to multiaddr
            if remote_address:
                observed_multiaddr = _remote_address_to_multiaddr(remote_address)
            else:
                observed_multiaddr = None
            logger.debug(
                "Connection from remote peer %s, address: %s, multiaddr: %s",
                peer_id,
                remote_address,
                observed_multiaddr,
            )
        except Exception as e:
            logger.error("Error getting remote address: %s", e)
            remote_address = None

        protobuf = _mk_identify_protobuf(host, observed_multiaddr)
        response = protobuf.SerializeToString()
        
        # Send the response
        try:
            await stream.write(response)
        except StreamClosed:
            logger.debug("Failed to respond to identify request: stream closed")
        else:
            await stream.close()
            logger.debug("Successfully handled identify request from %s", peer_id)

    return handle_identify
```

# Address Chain of Responsibility  
![ChainOfResponsibility](https://github.com/user-attachments/assets/dc530b95-ad88-4b88-b59b-72b748a55ea1)

The implementation uses a delegation pattern to retrieve remote addresses through the entire connection stack:  

1. `NetStream.get_remote_address()` → delegates to `MplexStream`  
2. `MplexStream.get_remote_address()` → delegates to `Mplex`  
3. `Mplex.get_remote_address()` → delegates to `SecureSession`  
4. `SecureSession.get_remote_address()` → delegates to `RawConnection`  
5. `RawConnection.get_remote_address()` → delegates to `TrioTCPStream`  
6. `TrioTCPStream.get_remote_address()` → retrieves the actual address from `socket.getpeername()`  

This chain of responsibility allows any component to access the remote address without needing to know the entire stack implementation details.  

## Address Conversion  

Socket addresses are converted to multiaddresses for use in the protocol:  

```python
def _remote_address_to_multiaddr(address: tuple[str, int]) -> Multiaddr:
    """Convert a (host, port) tuple to a multiaddr."""
    host, port = address
    # Detect if the address is IPv4 or IPv6
    try:
        socket.inet_pton(socket.AF_INET, host)  # IPv4
        return Multiaddr(f"/ip4/{host}/tcp/{port}")
    except (OSError, socket.error):
        try:
            socket.inet_pton(socket.AF_INET6, host)  # IPv6
            return Multiaddr(f"/ip6/{host}/tcp/{port}")
        except (OSError, socket.error):
            # Handle hostname as fallback
            return Multiaddr(f"/dns/{host}/tcp/{port}")
```


## Usage  

No explicit API changes are needed to use this feature. All peers running this version will automatically:  

- Report observed addresses back to peers that connect to them  


This information can be used for better connection management, address advertisement, and NAT traversal.  

## Future Work  

- Consider caching observed addresses in the `PeerStore` for faster reconnection  
- Extend to support other transport types beyond TCP  
- Would be nice to have a connection implementation like in js-libp2p that implements a connection "panel" info
```typescript
interface ConnectionInit {
  remoteAddr: Multiaddr
  remotePeer: PeerId
  newStream(protocols: string[], options?: AbortOptions): Promise<Stream>
  close(options?: AbortOptions): Promise<void>
  abort(err: Error): void
  getStreams(): Stream[]
  status: ConnectionStatus
  direction: Direction
  timeline: ConnectionTimeline
  multiplexer?: string
  encryption?: string
  limits?: ConnectionLimits
  logger: ComponentLogger
}

```

The current implementation provides the foundation for more advanced address discovery and NAT traversal techniques in future updates.

acul71 committed:
test: add test_complete_remote_address_delegation_chain

acul71 commented:
@pacrob @seetadev 
# Test Analysis: Remote Address Delegation Chain in libp2p

## Overview
This test https://github.com/libp2p/py-libp2p/pull/506/commits/e36b3e59778b3a43a7f1faba601a5a7ac1b63153 verifies the proper delegation of remote addresses through the entire libp2p network stack, ensuring address consistency across all protocol layers.

## Test Description
`test_complete_remote_address_delegation_chain` establishes a connection between two libp2p hosts and examines how the remote address is passed through each layer of the networking stack.

## Key Components Tested
- NetStream
- MplexStream
- Mplex
- SecureSession
- RawConnection
- TrioTCPStream
- Socket

## Test Flow
1. Creates two libp2p hosts (host_a and host_b)
2. Establishes a stream from host_a to host_b using the identify protocol
3. Retrieves references to components throughout the connection stack
4. Extracts the remote address from each layer
5. Verifies address consistency across all layers
6. Confirms the retrieved address matches host_b's actual address

## Significance
This test ensures the address delegation system functions correctly, which is essential for proper peer identification, connection management, and network reliability in the libp2p ecosystem.
